### PR TITLE
`Replicate` integration and remote callbacks

### PR DESCRIPTION
**Pull request by nicktrn**
## Description

This adds a Replicate integration and support for tasks with [remote callbacks](https://github.com/triggerdotdev/trigger.dev/discussions/502).

## Testing

See `references/job-catalog`

## Changelog

- Support tasks with remote callbacks
- Inherit common tsconfig for integrations
- Add Replicate integration
- Add pagination helpers
- Use inferred types for all `io.runTask()` calls

## Todo
- [x] Remote callbacks
- [x] Replicate tasks
- [x] Replicate helpers
- [x] Fail with error on timeout
- [x] Replicate docs
- [x] Update `io.runTask()` docs
- [x] More job-catalog entries
- [x] Sensible `io.runTask()` return type
- [x] Mirror final changes to `performRunExecutionV1`
- [x] Add changeset

### CODE REVIEW

File: apps/webapp/app/routes/api.v1.runs.$runId.tasks.$id.callback.$secret.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  const { runId, id } = ParamsSchema.parse(params);
+
+  // Now parse the request body
+  const anyBody = await request.json();
+
+  // Allows any valid object
+  // TODO: maybe add proper schema parsing during io.runTask(), or even skip this step
+  const body = CallbackTaskBodyInputSchema.safeParse(anyBody);
```
Comment (nicktrn) [inline apps/webapp/app/routes/api.v1.runs.$runId.tasks.$id.callback.$secret.ts]: Currently `z.object({}).passthrough()`. Leave as is? Change to `z.any()` or even remove schema parsing? There are no user-provided schemas for remote callbacks (yet), and we couldn't use them here.

File: packages/trigger-sdk/src/io.ts [hunk 2]
```diff
@@ -621,6 +605,11 @@ export class IO {
           task,
         });
 
+        // TODO: empty return? maybe don't even parse first
+        if (task.status === "WAITING" && task.callbackUrl) {
+          return output;
+        }
```
Comment (nicktrn) [inline packages/trigger-sdk/src/io.ts]: ~~Ideas? Gets discarded anyway.~~

_edit:_ Will return prior to parsing. Throwing immediately after.

**Comment by nicktrn** [comment]
@ericallam Take it apart :scissors: 

Didn't find a clean way to narrow the callbackUrl type, so took out the workaround in the end.

Same goes for the `io.runTask()` return, but will just add some docs re type casting the async return for now.

File: apps/webapp/app/routes/api.v1.runs.$runId.tasks.$id.callback.$secret.ts (hunk 1 above)
Comment (ericallam) [reply apps/webapp/app/routes/api.v1.runs.$runId.tasks.$id.callback.$secret.ts]: I think we can probably skip the schema parsing and just save the body to the task output without it

File: apps/webapp/app/routes/api.v1.runs.$runId.tasks.ts [hunk 3]
```diff
@@ -217,6 +228,15 @@ export class RunTaskService {
...
           { tx, runAt: task.delayUntil ?? undefined }
         );
+      } else if (task.status === "WAITING" && callbackUrl && taskBody.callback) {
+        // We need to schedule the callback timeout
+        await workerQueue.enqueue(
+          "processCallbackTimeout",
+          {
+            id: task.id,
+          },
+          { tx, runAt: new Date(Date.now() + taskBody.callback.timeoutInSeconds * 1000) }
```
Comment (ericallam) [inline apps/webapp/app/routes/api.v1.runs.$runId.tasks.ts]: What happens if the timeout isn't set? Is it required? Might want to only schedule this if the timeout is set and larger than 0

Comment (nicktrn) [reply apps/webapp/app/routes/api.v1.runs.$runId.tasks.ts]: The timeout currently defaults to an hour

File: apps/webapp/app/routes/api.v1.runs.$runId.tasks.ts [hunk 4]
```diff
@@ -233,16 +238,21 @@ export class RunTaskService {
...
+          delayUntilInFuture || callbackEnabled || taskBody.trigger
             ? "WAITING"
             : taskBody.noop
             ? "COMPLETED"
             : "RUNNING";
       }
 
+      const taskId = ulid();
+      const callbackUrl = callbackEnabled
+        ? `${env.APP_ORIGIN}/api/v1/runs/${runId}/tasks/${taskId}/callback/${generateSecret()}`
```
Comment (ericallam) [inline apps/webapp/app/routes/api.v1.runs.$runId.tasks.ts]: Could you update the `generateSecret` util function to receive a param with number of bytes (defaulted to 32) and pass in 12 here? The URLs are a little long using a 32 byte hex secret.

File: integrations/replicate/src/predictions.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+          ...params,
+          webhook: task.callbackUrl ?? "",
+          webhook_events_filter: ["completed"],
+        });
+      },
+      {
+        name: "Create And Await Prediction",
+        params,
+        properties: [...createPredictionProperties(params), ...callbackProperties(params)],
+        callback: { enabled: true },
```
Comment (ericallam) [inline integrations/replicate/src/predictions.ts]: These have default timeouts right?

Comment (nicktrn) [reply integrations/replicate/src/predictions.ts]: Yes

File: docs/sdk/io/runtask.mdx [hunk 6]
```diff
@@ -201,4 +219,43 @@ client.defineJob({
...
+```typescript Remote callbacks
+client.defineJob({
+  id: "remote-callback-example",
+  name: "Remote Callback example",
+  version: "0.1.1",
+  trigger: eventTrigger({ name: "predict" }),
+  integrations: { replicate },
+  run: async (payload, io, ctx) => {
+    //runTask
+    const prediction = await io.runTask(
```
Comment (ericallam) [inline docs/sdk/io/runtask.mdx]: I think this is meant to be `io.replicate.runTask` right?

Comment (nicktrn) [reply docs/sdk/io/runtask.mdx]: Yes, good catch!

Comment (nicktrn) [reply docs/sdk/io/runtask.mdx]: Had to update the GitHub example too, as underlying clients are not exposed at `io.integration.client` anymore. Also added a section to mention the wrappers at `io.integration.runTask()`.

File: docs/sdk/io/runtask.mdx [hunk 7]
```diff
@@ -6,6 +6,8 @@ description: "`io.runTask()` allows you to run a [Task](/documentation/concepts/
 
 A Task is a resumable unit of a Run that can be retried, resumed and is logged. [Integrations](/integrations) use Tasks internally to perform their actions.
 
+The wrappers at `io.integration.runTask()` expose the underlying Integration client as the first callback parameter (see examples on the right). They will have defaults set for options and `onError` handlers, but should otherwise be considered identical to raw `io.runTask()`.
+
```
Comment (nicktrn) [inline docs/sdk/io/runtask.mdx:10]: New section on integration `io.runTask()` wrappers

File: docs/sdk/io/runtask.mdx [hunk 8]
```diff
@@ -150,11 +170,11 @@ client.defineJob({
...
   run: async (payload, io, ctx) => {
     //runTask
-    const response = await io.runTask(
+    const response = await io.github.runTask(
       "create-card",
-      async () => {
+      async (client) => {
         //create a project card using the underlying GitHub Integration client
-        return io.github.client.rest.projects.createCard({
+        return client.rest.projects.createCard({
```
Comment (nicktrn) [inline docs/sdk/io/runtask.mdx:177]: Updated `io.runTask()` example

**Review by nicktrn — COMMENTED** [review]
Added a section on runTask wrappers as we have to make use of them now in the examples. (Underlying clients are not directly exposed anymore)

@matt-aitken You may want to have a brief look at this. Must be a remnant from the last big integration update.

**Review by ericallam — APPROVED** [review]

**Pull request merged**